staging/src/github.com/kubeedge/beehive/pkg/core/core.go

```diff
@@ -84,3 +90,46 @@ func moduleKeeper(name string, moduleInfo *ModuleInfo, m common.ModuleInfo) {
 		beehiveContext.AddModuleGroup(name, moduleInfo.module.Group())
 	}
 }
+
+// localModuleKeeper starts and tries to keep module running when module exited.
+// Call EnableModuleRestart() to enable auto-restarting feature in alpha version.
+func localModuleKeeper(m *ModuleInfo) {
+	if !moduleRestartEnabled {
+		m.module.Start()
+		return
+	}
+
+	ctx := beehiveContext.GetContext()
+	backoffDuration := time.Second
+
+	// do if module exits
+	afterFunc := func() {
+		if r := recover(); r != nil {
+			klog.Errorf("module %s panicking: %v", m.module.Name(), r)
+		}
+		m.active = false
+		klog.Errorf("module %s exited, will restart in %ds", m.module.Name(), int(backoffDuration.Seconds()))
+	}
+
+	for {
+		func() {
+			defer afterFunc()
+			m.active = true
+			m.module.Start()
```

Comment:
I'm a little confused here. The function here seems to be to start a module. If the startup fails, afterFunc will be executed and an error message will be output. But if the module is successfully started, will afterFunc still be executed? But there seems to be no handling of successful startup in afterFunc. Did I understand something wrong?

Reply:
Currently we are focusing on `EdgeCore` and this feature could only be implemented for edge-site modules so we care less about `CloudCore`. 

On edge site, the `Start()` function of `devicetwin`, `edged`, `edgehub`, `edgestream`, `eventbus`, `servicebus` will loop forever. Specially, `metamanager` starts two goroutine and returns. But the function `runMetaManager()` tells that it is Main loop at L425, `metamanager/process.go`. So this PR treats it as the main loop of the module and removes the `go func`.

Ideally, if a goroutine is started by some module, the module does have responsibility to handle the goroutine to prevent producing orphans. But it need plenty of work to modify all modules, especially for `CloudCore`. 

With restarting of cloud-site modules disabled, it works as normal and `afterFunc` will not be called, which post-processes module exiting.

In this PR, cloud-site restarting cannot be enabled, so it will not ruin cloud-site logic for starting modules. To make it further, maybe a `<-beehiveContext.Done()` could be added to `Start()` to keep a main goroutine of modules.